src/doc/contrib/src/tests/writing.md

```diff
@@ -6,65 +6,69 @@ tests is also encouraged!
+
+There are two styles of tests that can roughly be categorized as
+- functional tests
+  - The fixture is programmatically defined
+  - The assertions are regular string comparisons
+  - Easier to share in an issue as a code block is completely self-contained
+  - More resilient to insignificant changes though ui tests are easy to update when a change does occur
+- ui tests
+  - The fixture is file-based
+  - The assertions use file-backed snapshots that can be updated with an env variable
+  - Easier to review the expected behavior of the command as more details are included
+  - Easier to get up and running from an existing project
+  - Easier to reason about as everything is just files in the repo
 
 These tests typically work by creating a temporary "project" with a
 `Cargo.toml` file, executing the `cargo` binary process, and checking the
 stdout and stderr output against the expected output.
 
-### `cargo_test` attribute
-
-Cargo's tests use the `#[cargo_test]` attribute instead of `#[test]`. This
-attribute injects some code which does some setup before starting the test,
-creating the little "sandbox" described below.
-
-### Basic test structure
-
-The general form of a test involves creating a "project", running `cargo`, and
-checking the result. Projects are created with the [`ProjectBuilder`] where
-you specify some files to create. The general form looks like this:
-
-```rust,ignore
-let p = project()
-    .file("src/main.rs", r#"fn main() { println!("hi!"); }"#)
-    .build();
-```
-
-The project creates a mini sandbox under the "cargo integration test"
-directory with each test getting a separate directory such as
-`/path/to/cargo/target/cit/t123/`. Each project appears as a separate
-directory. There is also an empty `home` directory created that will be used
-as a home directory instead of your normal home directory.
-
-If you do not specify a `Cargo.toml` manifest using `file()`, one is
-automatically created with a project name of `foo` using `basic_manifest()`.
-
-To run Cargo, call the `cargo` method and make assertions on the execution:
+### Functional Tests
 
+Generally, a functional test will be placed in `tests/testsuite/<command>.rs` and will look roughly like:
 ```rust,ignore
-p.cargo("run --bin foo")
-    .with_stderr(
-        "\
-[COMPILING] foo [..]
-[FINISHED] [..]
-[RUNNING] `target/debug/foo`
-",
-    )
-    .with_stdout("hi!")
-    .run();
+#[cargo_test]
+fn <description>() {
+    let p = project()
+        .file("src/main.rs", r#"fn main() { println!("hi!"); }"#)
+        .build();
+
+    p.cargo("run --bin foo")
+        .with_stderr(
+            "\
+    [COMPILING] foo [..]
+    [FINISHED] [..]
+    [RUNNING] `target/debug/foo`
+    ",
+        )
+        .with_stdout("hi!")
+        .run();
+    }
+}
 ```
 
-This uses the [`Execs`] struct to build up a command to execute, along with
-the expected output.
+`cargo_test` attribute:
```

Comment:
> Why was this moved under Functional Tests?
> ...
> ...it might be better to have it as its own section.

For me, it fit well to enumerate all the parts of a test style under itself as they are very different from each other.  This keeps a coherent flow to the information and all relevant information stays close to each other.  Yes, it means a topic is covered twice but there are other care abouts to balance with DRY.

Reply:
That makes sense to me!